esrally/telemetry.py

```diff
@@ -150,26 +150,49 @@ class InternalTelemetryDevice(TelemetryDevice):
     internal = True
 
 
-class SamplerThread(threading.Thread):
-    def __init__(self, recorder):
-        threading.Thread.__init__(self)
+class Sampler:
+    """This class contains the actual logic of SamplerThread for unit test purposes."""
+
+    def __init__(self, recorder, *, sleep=time.sleep):
         self.stop = False
         self.recorder = recorder
+        self.sleep = sleep
 
     def finish(self):
         self.stop = True
-        self.join()
 
     def run(self):
         # noinspection PyBroadException
         try:
-            while not self.stop:
-                self.recorder.record()
-                time.sleep(self.recorder.sample_interval)
+            sleep_left = self.recorder.sample_interval
+            while True:
+                if sleep_left <= 0:
+                    self.recorder.record()
+                    sleep_left = self.recorder.sample_interval
+
+                if self.stop:
+                    break
+
+                sleep_seconds = sleep_left
```

Comment:
may I suggest the following instead of this block, with the explainer comment:
```suggestion
                # check for self.stop at least every second
                sleep_seconds = min(sleep_left, 1)
```

Reply:
That's much better, thanks!